Some decently working field split options

lindsayad opened the pull request:
Performance is on par with LU up to about 350k dofs

To-do:

- [ ] Test options for a variety of Reynolds numbers. E.g. likely AMG will perform poorly for high Reynolds numbers

Refs #22359

@GiudGiud @tanoret let's continue working to figure out a "best" (acknowledging that best is likely fluid and dependent on the problem/operator) set of options that we can put forward as an example for users

lindsayad committed:
Some decently working field split options

Performance is on par with LU up to about 350k dofs

Refs #22359

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-no-slip.i:
```diff
@@ -131,11 +131,75 @@ velocity_interp_method = 'rc'
+      # is not diagonally dominant. CAN NOT use additive, multiplicative and etc.
+      # Original system:
+      # | A B | | u | = | f_u |
+      # | C 0 | | p |   | f_v |
```
dont we have some pressure terms in the mass equation thx to RC?

lindsayad replied:
we do, but they are small compared to the velocity. This is a typical block
```
row 0: (0, 8.25)  (1, 0.55)  (2, -0.959416)
row 1: (0, 0.55)  (1, 8.25)  (2, -0.959416)
row 2: (0, 0.55)  (1, 0.55)  (2, 0.0405844)
```
where 0 and 1 are velocity dofs and 2 is a pressure dof

GiudGiud replied:
how about ~0 in the matrix then?

lindsayad replied:
I don't understand the question

GiudGiud replied:
instead of saying A B C and 0
write
A B C and ~0

lindsayad replied:
Lol I finally understand that your question was directly about this input file comment. Yes I will make sure this comment gets changed

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1202788/) on d894f4a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22405/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1202801/) on d894f4a wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22405/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
what kind of documentation would we want accompanying this?
Some paragraph somewhere in NSFV about solver schemes with a table, some runtimes, and some memory consumptions for a simple case?

lindsayad committed:
3D example

This uses the options Guillaume sent me that he and Fande worked on

lindsayad commented:
Ok with 3D, FSP using the options that @GiudGiud sent in slack looks *really* good compared to LU for just an 80k dof problem on 4 procs

![lu](https://user-images.githubusercontent.com/10248304/197086658-30a650cb-1e82-4212-8f27-b7f3d69f5668.png)
![fsp](https://user-images.githubusercontent.com/10248304/197086664-ce2837d8-ef95-441f-a1d8-8e59ad52b5d2.png)

lindsayad commented:
@oanaoana tagging you so that you are aware of any activity on this PR

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-no-slip.i:
```diff
@@ -131,11 +131,75 @@ velocity_interp_method = 'rc'
+      # PETSc options for this subsolver
+      # A prefix will be applied, so just put the options for this subsolver only
+      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type -ksp_rtol'
+      petsc_options_value = 'hypre    boomeramg      gmres     1e-4'
```
btw I m not sure if this is on purpose, but on the other field split we set it to precondition only, with gmres here as the ksp_type I think it solves it

lindsayad commented:
@andrsd pinging you just for fun. Feel free to unsubscribe 😄

lindsayad commented:
Noting that `-pc_fieldsplit_schur_precondition self` (which creates a `MATSCHURCOMPLEMENT` which only supports `-pc_type lsc` (and apparently not all the time given the errors I've gotten when trying it)) appears to yield a very powerful overall preconditioning affect, perhaps always stronger than `-pc_fieldsplit_schur_precondition selfp`

andrsd commented:
> @andrsd pinging you just for fun. Feel free to unsubscribe 😄

[PETSC example 46](https://petsc.org/release/src/ts/tutorials/ex46.c.html) (incompressible NS) may have some interesting bits for you. For one it has a manufactured solution that you could repurpose for your testing. I think I played with it a bit and with pure LU, the solver would not converge to the right solution. With field split it would. I am not sure why the LU failed (in my pre INL times we had a LU-based FEM solver for incompressible NS and all worked fine). It could have been a user error.

Another bit of (hopefully) useful information. When we had a junction model in RELAP-7 based on Mortar method, it generated a matrix with a similar structure. Using the PETSc LU solver, we did not converge to the right solution, but with MUMPS we would. Could be related to the above paragraph. We did not try SuperLU, IIRC.

lindsayad commented:
Yea a lot of constraints that mortar can tackle (like solution continuity, mechanical contact) generate saddle point problems like incompressible Navier-Stokes. Fascinating that LU would actually lead to a wrong solution as opposed to simply not converging

lindsayad commented:
Added to back of task list

lindsayad commented:
Oh ok you can run `-pc_fieldsplit_schur_precondition self` if you don't specify a `-pc_type` for the pressure block

lindsayad commented:
Some timings for 2D testing

## Re = 2.2

### Time = 17.694
```
      petsc_options_iname = '-pc_fieldsplit_schur_fact_type  -pc_fieldsplit_schur_precondition'
      petsc_options_value = 'full                            self'
    []
    [u]
      vars = 'vel_x vel_y'
      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type -ksp_rtol'
      petsc_options_value = 'hypre    boomeramg      gmres     1e-4'
    []
    [p]
      vars = 'pressure'
```
### Time = 15.258
```
      petsc_options_iname = '-pc_fieldsplit_schur_fact_type  -pc_fieldsplit_schur_precondition'
      petsc_options_value = 'full                            selfp'
    []
    [u]
      vars = 'vel_x vel_y'
      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type -ksp_rtol'
      petsc_options_value = 'hypre    boomeramg      gmres     1e-4'
    []
    [p]
      vars = 'pressure'
      petsc_options_iname = '-pc_type -ksp_type -ksp_rtol'
      petsc_options_value = 'jacobi   gmres     1e-4'
```
### Time = 27.896
```
      petsc_options_iname = '-pc_fieldsplit_schur_fact_type  -pc_fieldsplit_schur_precondition'
      petsc_options_value = 'full                            selfp'
    []
    [u]
      vars = 'vel_x vel_y'
      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type -ksp_rtol'
      petsc_options_value = 'hypre    boomeramg      gmres     1e-4'
    []
    [p]
      vars = 'pressure'
      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type -ksp_rtol'
      petsc_options_value = 'hypre    boomeramg      gmres     1e-4'
```
I thought I'd try multigrid in case the Schur complement is like an elliptic operator? Definitely doesn't seem to work well.
### Time = 30.121
```
      petsc_options_iname = '-pc_fieldsplit_schur_fact_type  -pc_fieldsplit_schur_precondition'
      petsc_options_value = 'full                            selfp'
    []
    [u]
      vars = 'vel_x vel_y'
      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type -ksp_rtol'
      petsc_options_value = 'hypre    boomeramg      gmres     1e-4'
    []
    [p]
      vars = 'pressure'
      petsc_options_iname = '-pc_type -pc_factor_mat_solver_type -ksp_type -ksp_rtol'
      petsc_options_value = 'lu       mumps                      gmres     1e-4'
```
### Time = 16.328
```
      petsc_options_iname = '-pc_fieldsplit_schur_fact_type  -pc_fieldsplit_schur_precondition'
      petsc_options_value = 'full                            selfp'
    []
    [u]
      vars = 'vel_x vel_y'
      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type'
      petsc_options_value = 'hypre    boomeramg      preonly'
    []
    [p]
      vars = 'pressure'
      petsc_options_iname = '-pc_type -pc_factor_mat_solver_type -ksp_type'
      petsc_options_value = 'lu       mumps                      preonly'
```
### Time = 17.265
```
      petsc_options_iname = '-pc_fieldsplit_schur_fact_type  -pc_fieldsplit_schur_precondition'
      petsc_options_value = 'full                            selfp'
    []
    [u]
      vars = 'vel_x vel_y'
      petsc_options_iname = '-pc_type -pc_factor_mat_solver_type -ksp_type'
      petsc_options_value = 'lu       mumps                      preonly'
    []
    [p]
      vars = 'pressure'
      petsc_options_iname = '-pc_type -pc_factor_mat_solver_type -ksp_type'
      petsc_options_value = 'lu       mumps                      preonly'
```

lindsayad commented:
Short summary:

| PC | time |
| --- | --- |
| hypre-g-self-g | 17.694 |
| hypre-g-jacobian-g | 15.258 |
| hypre-g-hypre-g | 27.896 |
| hypre-g-lu-g | 30.121 |
| hypre-p-lu-p | 16.328 |
| lu-p-lu-p | 17.275 |
| hypre-g-jacobi-p | 12.534 |
| hypre-g-ilu-p | 16.869 |
| hypre-p-jacobi-p | 9.954 |
| jacobi-p-jacobi-p | fail |
| jacobi-p-self-g | fail |

-

lindsayad commented:
Yea it is significantly better than LU on big problems for non-advection dominated flows. But I want to find a good option for advection dominated too. Maybe with viscosity due to turbulence it would be ok again.

I started out the day very happy with the low Re results and then ended depressed after working on the high Re cases. I’m currently looking into AIR for boomeramg which I hear is supposed to be something to try for advection dominant problems, but am hitting some floating point exceptions. Asked the PETSc list about it and they’ve suggested I forward my question onto the hypre team.

I am just really hoping there’s a good iterative PC for high Re NS; it seems like there’s got to be some really smart people who have studied this in detail. It just can’t be that the only answer is to treat the convection term explicitly. 

Honestly I need to look again at how OpenFOAM preconditions the momentum equation in SIMPLE. @grmnptr?

grmnptr commented:
I think a 121 s solve time for 1.5M dofs is pretty cool!

lindsayad commented:
How about a 43 second solve time for Re=2.2! 😄

lindsayad commented:
@prj- has given me some option suggestions for leveraging his pretty cool sounding domain decomposition method described in https://arxiv.org/pdf/2201.02250.pdf. I'm going to try giving that a whirl for Re=220

prj- commented:
Thanks for pinging me! Currently building MOOSE so that it will be easier for me to help you out.

prj- commented:
Oh, well... didn't request anything, yet got
```
[...]
<<< User requested thread model: openmp >>>
checking for OpenMP flag of C++ compiler... unknown
checking for OpenMP flag of C compiler... unknown
checking for OpenMP flag of Fortran 77 compiler... unknown
configure: error: requested openmp threading model, but C++ compiler does not support openmp.
Running make -j 1...
make: *** No targets specified and no makefile found.  Stop.
```

lindsayad commented:
Is this while configuring libMesh? If so, then you should add `--with-thread-model=none --disable-openmp`. The latter option may be unnecessary but I would add it to be safe

prj- commented:
What about
```
configure.ac:138: installing 'build-aux/compile'
configure.ac:65: installing 'build-aux/missing'
Makefile.am: installing 'build-aux/depcomp'
autoreconf: Leaving directory '.'
patching file 'build-aux/compile'
Reversed (or previously applied) patch detected!  Assume -R? [y] n
Apply anyway? [n] n
3 out of 3 hunks ignored--saving rejects to 'build-aux/compile.rej'
Bootstrapping using autoreconf -v -f -i ...
```
I don't want to flood the MR with unrelated stuff, so feel free to redirect me somewhere else. libMesh seems to be building for now.

lindsayad commented:
If it gets too much longer we can move it to a MOOSE discussion. Seems like maybe just a difference in autotools versions ... if it's building, then let's not worry too much about it

prj- commented:
Looks like libMesh (at least the snapshot used in MOOSE `master` branch -- should I try `next`?) is not quite capable of handling PETSc `main` branch `--with-strict-petscerrorcode`, which I use on all my `${PETSC_ARCH}`.
```
../src/base/libmesh.C:512:94: error: assigning to 'PetscErrorCode' from incompatible type 'int'
      ierr = DMRegister(DMLIBMESH, DMCreate_libMesh); CHKERRABORT(libMesh::GLOBAL_COMM_WORLD,ierr);
```
Too bad.

lindsayad commented:
> Looks like libMesh (at least the snapshot used in MOOSE `master` branch -- should I try `next`?)

We are fairly behind on our petsc submodule (3.16.5) although we are planning to update to 3.19 soon. So you'll get this failure whether you're `master`, `next`, etc.

lindsayad committed:
Apply best options found for Re=2.2

lindsayad committed:
Focus on 2D case to start

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1465745/) on e737ccc : invalidated by @lindsayad

moosebuild commented:
Job [Apptainer build apps](https://civet.inl.gov/job/1465725/) on e737ccc : invalidated by @lindsayad

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1465746/) on e737ccc : invalidated by @lindsayad

lindsayad commented:
`-pc_type` for pressure:
- jacobi: 39.833s (834 linear iterations)
- none: 41.255s (897 linear iterations)

GiudGiud commented:
Can you try hypre?

lindsayad commented:
It is kind of disingenuous to call it the `-pc_type` for pressure as it's a preconditioner on the Schur KSP. You can look at the timings table above. `hypre boomeramg` is really not good when applied to "pressure"

grmnptr commented:
When I wrote my DG Schur compelment based NS solver in deal.ii, I built a simple mass matrix on the pressure variable for preconditioning. It gave good results. there is also a tutorial about it somewhere.

lindsayad commented:
The issue I am seeing is in ksp(A00, Ap00) though (e.g. solving the velocity-velocity block)

lindsayad commented:
| Number of eigenvectors | total linear iterations |
| --- | --- |
| 10 | 190 |
| 20 | 194 |
| 30 | 171 |
| 40 | 176 |
| 80 | too slow |

lindsayad committed:
Testing HPDDM options for high Reynolds number case

lindsayad commented:
@oanaoana I've pushed the commit with the HPDDM options, so you are free to test/explore

lindsayad commented:
| inner ksp rtol | solve time |
| --- | --- |
| preonly | 70.479 |
| 0.5 | 62.865 |
| 0.1 | 97.952 |

lindsayad commented:
For the 220 Reynolds number: with boomeramg on u and Jacobi on the Schur, and with inner ksp solve tolerances of 0.1, then I get a solve of 25.807 seconds for the 1.5 million dof problem on rod with 64 processes and a solve time of 77.074 seconds with LU.

lindsayad committed:
Strong field split options for Reynolds number of 220

lindsayad committed:
Finalize field split options

Also:
- I don't love having the relatively large `abs_zero` for the field
  split test but I've inspected the output and it looks good

prj- commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-no-slip.i:
```diff
@@ -131,12 +134,86 @@ velocity_interp_method = 'rc'
+    [u]
+      vars = 'vel_x vel_y'
+      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type -ksp_rtol -ksp_gmres_restart'
+      petsc_options_value = 'hypre    boomeramg      fgmres    5e-1      300'
```
Is there a good reason to use FGMRES as an inner solver?

lindsayad replied:
I cannot answer you from the linear algebra perspective. As you know I didn't even know that it was invalid to be using gmres in the outer linear solve. But in my timings of my application, I see fgmres with the inner solves running faster than gmres with the inner solves.

| dofs | inner ksp solve type | time |
| --- | --- | --- |
| 384000 | fgmres | 19.392 |
| 384000 | gmres | 24.897 |
| 1536000 | fgmres | 25.598 |
| 1536000 | gmres | 35.807 |

prj- replied:
What about `-ksp_type gmres -ksp_pc_side right`?

prj- replied:
The difference you are seeing between `KSPFGMRES` versus `KSPGMRES` + `PC_RIGHT` is because the latter applies the `PC` an extra time, but it does not need to store an extra Krylov basis (which is the case of the former).

lindsayad replied:
I love getting your input on this PR. I've been learning so much!

prj- replied:
Flexible Krylov methods necessarily apply preconditioners on the right. You can't do `-ksp_type fgmres -ksp_pc_side left`, that will throw an error. You can do `-ksp_type fgmres -ksp_pc_side right`, but that's just redundant.

tanoret commented:
Results for MCRE-like reactor. With LU 837.60 s with latest prec by @lindsayad 437.94 s

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1470887/) on 30392a3 : invalidated by @lindsayad

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-no-slip.i:
```diff
@@ -131,12 +134,86 @@ velocity_interp_method = 'rc'
 []
 
+[Preconditioning]
+  active = SMP
```
```suggestion
  active = FSP
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/tests:
```diff
@@ -33,14 +33,28 @@
+      abs_zero = 1e-9
+      valgrind = HEAVY
+      detail = 'monolithic preconditioning technique'
+    []
```
```suggestion
       cli_args = 'Preconditioning/active=SMP'
    []
```

lindsayad committed:
Make FSP the default for this input

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

GiudGiud reviewed (dismissed):
awesome

lindsayad committed:
Run tris test with SMP

In order to avoid having to fiddle with `abs_zero`

TLWise commented:
Is there an example of this method using FEM rather than FVM?

GiudGiud commented:
There is a single one, in this file:
modules/navier_stokes/test/tests/finite_element/ins/pressure_channel/open_bc_pressure_BC_fieldSplit.

lindsayad commented:
Not yet, but there will be. I have been doing a lot of research on field split in the past couple weeks and been making some very good progress. One of the hardest problems is preconditioning the Schur complement. I've done some PETSc development in [this MR](https://gitlab.com/petsc/petsc/-/merge_requests/6642) and now I'm seeing mesh independent scaling of the iteration count for least squares commutator preconditioning which is pretty exciting.
